hub/policy/src/test/java/uk/gov/ida/hub/policy/domain/state/serialization/DeserializationTest.java

```diff
@@ -0,0 +1,296 @@
+import static uk.gov.ida.hub.policy.builder.state.IdpSelectedStateBuilder.anIdpSelectedState;
+import static uk.gov.ida.hub.policy.builder.state.MatchingServiceRequestErrorStateBuilder.aMatchingServiceRequestErrorState;
+import static uk.gov.ida.hub.policy.builder.state.NoMatchStateBuilder.aNoMatchState;
+import static uk.gov.ida.hub.policy.builder.state.PausedRegistrationStateBuilder.aPausedRegistrationState;
+import static uk.gov.ida.hub.policy.builder.state.RequesterErrorStateBuilder.aRequesterErrorState;
+import static uk.gov.ida.hub.policy.builder.state.SessionStartedStateBuilder.aSessionStartedState;
+import static uk.gov.ida.hub.policy.builder.state.SuccessfulMatchStateBuilder.aSuccessfulMatchState;
+import static uk.gov.ida.hub.policy.builder.state.TimeoutStateBuilder.aTimeoutState;
+import static uk.gov.ida.hub.policy.builder.state.UserAccountCreatedStateBuilder.aUserAccountCreatedState;
+import static uk.gov.ida.hub.policy.builder.state.UserAccountCreationFailedStateBuilder.aUserAccountCreationFailedState;
+import static uk.gov.ida.hub.policy.builder.state.UserAccountCreationRequestSentStateBuilder.aUserAccountCreationRequestSentState;
+import static uk.gov.ida.hub.policy.domain.state.serialization.SerializationHelpers.assertDeserializedForm;
+
+/**
+ * Tests that check that the "State" objects which we store in infinispan can be deserialized.
+ *
+ * In each case we take a blob of base64 encoded serialized data, deserialize it, and check that it matches an expected
+ * object (by comparing field-by-field).
+ *
+ * These tests are intended to catch the situation where a change is made to a State that means serialized data cannot be
+ * deserialized cleanly to the state objects. This usually causes issues during deployment, since there are existing States
+ * in the data store which the new code cannot handle.
+ *
+ * If you change a State in a way that breaks these tests you should first see if your change is actually breaking, or
+ * if there's something else going on (e.g. changing the default in a builder).
+ *
+ * If it looks like your changes are breaking you should see if you can make the changes in a non-breaking way (e.g. by
+ * marking new fields that don't need to be persisted as "transient", or by implementing
+ * {@link java.io.Externalizable})
+ *
+ * If you can't make the changes in a non-breaking way you should:
+ * - Ensure that deploying the code will not cause an outage, or in extreme cases agree a planned outage with stakeholders
+ * - Increment the serialVersionUID on the class you're breaking
+ * - Update the serialized data in the test so the test covers the new format
+ *
+ * @see <a href="https://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html">The javadoc for Serializable</a>
+ * @see <a href="https://stackoverflow.com/a/285809/1344760">Stackoverflow explanation of serialVersionUID</a>
+ */
+public class DeserializationTest {
```

Comment:
Is there possibly a test we can write that checks whether we're testing a State class against its deserialized form? Perhaps something narly with reflection?

Reply:
I thought about doing that, yeah. It would be possible, but it gets a bit filthy-looking. You'd have to:

* Use reflection to get all of the non-abstract classes implementing `State`
* Use reflection to list all of the methods on DeserializationTest
* Check that there's a correctly named method for every state

As it is there's a risk someone will add a new state, and these tests will give us a false sense of security, so I think that does have some value. If I get a chance once I've finished my current task I'll have a look at that. I'd prefer it if it didn't block this PR from getting merged though.

Reply:
Turns out `Use reflection to get all of the non-abstract classes implementing State` is actually pretty hard, due to java. Not sure it's worth it.